astropy/utils/misc.py

```diff
@@ -1159,6 +1161,27 @@ def dtype_bytes_or_chars(dtype):
     return out
 
 
+def unbroadcast(array):
+    """
+    Given an array, return a new array that is the smallest subset of the
+    original array that can be re-broadcasted back to the original array.
+
+    See https://stackoverflow.com/questions/40845769/un-broadcasting-numpy-arrays
+    for more details.
+    """
+
+    if array.ndim == 0:
+        return array
+
+    new_shape = np.where(np.array(array.strides) == 0, 1, array.shape)
```

Comment:
Hmm, it may be that the numpy version is one where there was a bug on interpreting writable; I certainly cannot reproduce it on latest numpy. But how about the following replacement?
```
array = array[tuple((slice(0, 1) if stride == 0 else slice(None)) for stride in array.strides)]
```
(I think this may even be faster...) I'll just push it up for speed...